### PR TITLE
only show directly inherited types in the InheritedBy progression query

### PR DESCRIPTION
**Pull request by brettfo**
Fixes internal bug 747965.

When showing the derived types in a progression query in the Solution Explorer, only the types that immediately derive from the selected type should be shown, not everything down the chain (because the tree view will take care of that).  This takes the behavior back to how it was in Dev12.

### CODE REVIEW

**Comment by brettfo** [comment]
Tagging @Pilchie @dpoeschl @jasonmalinowski @DustinCampbell @rchande @balajikris @jmarolf @davkean as potential reviewers.

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs [hunk 1]
```diff
@@ -32,7 +27,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
                 if (namedType.TypeKind == TypeKind.Class)
                 {
                     var derivedTypes = await namedType.FindDerivedClassesAsync(solution, cancellationToken).ConfigureAwait(false);
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs]: Doesn't this does a bunch of extra work to go and find all the levels of inheritance, only to ignore it after you have it?  Consider making `FindDerivedClassesAsync` able to return just one level directly.

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs [hunk 2]
```diff
@@ -32,7 +27,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
...
-                    foreach (var derivedType in derivedTypes)
+                    foreach (var derivedType in derivedTypes.Where(derived => derived.BaseType == namedType))
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs]: We don't have another helper for this somewhere? I feel bad we're doing lots of work chasing types only to throw them out.

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs [hunk 3]
```diff
@@ -41,7 +36,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
                 else if (namedType.TypeKind == TypeKind.Interface)
                 {
                     var derivedTypes = await namedType.FindDerivedInterfacesAsync(solution, cancellationToken).ConfigureAwait(false);
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs]: Unrelated to your change, but in Dev12 the Derived Types query we used to include classes that implemented the interface, but in the same hierarchical fashion that you're doing here now (as opposed to the new Implemented By query where flatten the list of implementers). Let's consider bringing that back.

**Comment by brettfo** [comment]
I've updated the helpers (only called by this progression query) to filter types when searching and not after, so we avoid doing extra work.  I've also added more to the tests that verify we brought the behavior back in line with Dev12 as mentioned in @dpoeschl's comment around showing types that directly implement an interface.

**Comment by Pilchie** [comment]
:+1:

File: src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb [hunk 4]
```diff
@@ -74,10 +78,16 @@ using System;
 
 public interface $$I { }
 
+public class C : I { } // should appear as being derived from (implementing) I
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb]: Do you have/need a case for basically ```class C2 : C {} // should not appear as being derived from I```?

Comment (brettfo) [inline src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb]: Good catch, I'll add that.

File: src/Workspaces/Core/Portable/Shared/Extensions/INamedTypeSymbolExtensions.cs [hunk 5]
```diff
@@ -25,12 +25,12 @@ public static IEnumerable<INamedTypeSymbol> GetBaseTypesAndThis(this INamedTypeS
...
         }
 
-        public static Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
+        public static Task<IEnumerable<INamedTypeSymbol>> FindTypesImmediatelyDerivedFromClassesAsync(
             this INamedTypeSymbol type,
             Solution solution,
             CancellationToken cancellationToken)
         {
-            return FindDerivedClassesAsync(type, solution, null, cancellationToken);
+            return DependentTypeFinder.GetTypesImmediatelyDerivedFromClassesAsync(type, solution, cancellationToken);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Extensions/INamedTypeSymbolExtensions.cs]: Nitpick: can we just get rid of extension methods like this that just forward to a static somewhere? It just makes the codebase harder to navigate.

File: src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb [hunk 6]
```diff
@@ -74,10 +78,18 @@ using System;
...
+
+public class C2 : C { } // should not appear as being derived from (implementing ) I
+
 interface I2 : I, IComparable
 {
     void M();
 }
+
+interface I3 : I2 // should not be shown as inherited by I
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb]: Can we have a VB test for these as well? VB's syntax at least treats these cases as different, I forget if the symbol model does.

**Comment by jasonmalinowski** [comment]
:+1:, other than the build failing and nitpicks. :-)

**Comment by brettfo** [comment]
Added VB tests and removed single-use extension methods to directly call the static helpers instead.

**Pull request merged**